File: coffee-backend/app/schemas/social.py
"""Schemas for Social Contacts & Sharing (friends, groups, share targets)."""
from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


def _initials(nome: str) -> str:
    """Generate initials from name (e.g. 'Ana Beatriz' -> 'AB')."""
    parts = nome.strip().split()
    if len(parts) >= 2:
        return (parts[0][0] + parts[1][0]).upper()
    elif parts:
        return parts[0][:2].upper()
    return "?"


# ── Friends ───────────────────────────────────────────────────────────────────

class SendFriendRequest(BaseModel):
    addressee_email: str | None = None
    addressee_id: UUID | None = None


class FriendResponse(BaseModel):
    id: UUID
    user_id: UUID
    nome: str
    email: str
    initials: str
    status: str
    created_at: datetime


class UserSearchResult(BaseModel):
    id: UUID
    nome: str
    email: str
    initials: str
    is_friend: bool
    friendship_status: Optional[str] = None


# ── Groups ────────────────────────────────────────────────────────────────────

class CreateGroupRequest(BaseModel):
    nome: str = Field(min_length=1, max_length=100)
    member_ids: list[UUID] = []


class AddMemberRequest(BaseModel):
    user_id: UUID


class GroupMemberResponse(BaseModel):
    user_id: UUID
    nome: str
    initials: str
    role: str


class GroupResponse(BaseModel):
    id: UUID
    nome: str
    is_auto: bool
    turma: Optional[str] = None
    disciplina_id: Optional[UUID] = None
    member_count: int
    members: Optional[list[GroupMemberResponse]] = None
    created_at: datetime


# ── Share Targets ─────────────────────────────────────────────────────────────

class ShareTargetResponse(BaseModel):
    friends: list[FriendResponse]
    groups: list[GroupResponse]

File: coffee-frontend/Models/Social.swift
import Foundation

// MARK: - Friend

struct Friend: Codable, Identifiable {
    let id: String
    let userId: String
    let nome: String
    let email: String
    let initials: String
    let status: FriendStatus
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id, nome, email, initials, status
        case userId = "user_id"
        case createdAt = "created_at"
    }
}

enum FriendStatus: String, Codable {
    case pendingSent = "pending_sent"
    case pendingReceived = "pending_received"
    case accepted
}

// MARK: - Group

struct SocialGroup: Codable, Identifiable {
    let id: String
    let nome: String
    let isAuto: Bool
    let disciplinaId: String?
    let memberCount: Int
    let members: [GroupMember]?
    let createdAt: Date?

    enum CodingKeys: String, CodingKey {
        case id, nome, members
        case isAuto = "is_auto"
        case disciplinaId = "disciplina_id"
        case memberCount = "member_count"
        case createdAt = "created_at"
    }
}

struct GroupMember: Codable, Identifiable {
    var id: String { userId }
    let userId: String
    let nome: String
    let initials: String
    let role: String

    enum CodingKeys: String, CodingKey {
        case nome, initials, role
        case userId = "user_id"
    }
}

// MARK: - User Search Result

struct UserSearchResult: Codable, Identifiable {
    let id: String
    let nome: String
    let email: String
    let initials: String
    let isFriend: Bool
    let friendshipStatus: String?

    enum CodingKeys: String, CodingKey {
        case id, nome, email, initials
        case isFriend = "is_friend"
        case friendshipStatus = "friendship_status"
    }
}

// MARK: - Share Targets

struct ShareTargets: Codable {
    let friends: [Friend]
    let groups: [SocialGroup]
}

// MARK: - Requests

struct SendFriendRequestBody: Codable {
    let addresseeEmail: String?
    let addresseeId: String?
    enum CodingKeys: String, CodingKey {
        case addresseeEmail = "addressee_email"
        case addresseeId = "addressee_id"
    }
}

struct CreateGroupRequest: Codable {
    let nome: String
    let memberIds: [String]
    enum CodingKeys: String, CodingKey {
        case nome
        case memberIds = "member_ids"
    }
}

struct AddMemberRequestBody: Codable {
    let userId: String
    enum CodingKeys: String, CodingKey {
        case userId = "user_id"
    }
}
